Feature screen three domain

anarahari opened the pull request:
- Created EpisodeRepository interface to get episode details

- Created GetEpisodeDetailsUseCase to fetch the episode details response from repository

- Added unit test cases for GetEpisodeDetailsUseCase to fetch episode details on success and failure scenarios.

- Added characters variable to Episode model to fetch list of characters.

- Changes for lintchecks

anarahari committed:
feat : Screen Two Domain

- added GetCharacterDetailsUseCase
- added getCharacterDetails function in character repository

anarahari committed:
feat : Screen Two Domain

- lint checks resolved

anarahari committed:
feat : Screen Two Domain

- CharacterModel updated with episode and origin model fields

anarahari committed:
feat : Screen Two Domain

- repo impl function for getCharacterDetails

anarahari committed:
feat : Second Screen Data Layer

- Added graphql for GetCharacterDetails
- Repo impl for GetCharacterDetails
- Added DI for GetCharacterDetailsUseCase in DataModule
- added extension functions in data mapper for origin model and episode model

anarahari committed:
feat : Second Screen Data Layer

- main.yaml changes to run github actions for all branches

anarahari committed:
feat : Second Screen Data Layer

- ktlint checks

anarahari committed:
feat : Feature screen two presentation

- Added Nav graph, nav controller for presentation layer
- Added Navroutes for both screens for routing purpose

anarahari committed:
feat : Feature screen two presentation

- Composable functions to display Character Details Screen starting with Image, Character details, Location, Origin and Episodes

- First Screen : Added LaunchedEffect for api call to get characters

- First Screen - Added navigation to character details on click of list item

-  Custom App Bar : dynamic title

- Dimens.Kt file to display custom sizes of text, paddings

- CharacterModel : Added Location model

- Passed GetCharacterDetailsUseCase as a parameter to viewmodelprovider

- DataMapper : Added location model mapper

- Lintcheck changes

anarahari committed:
feat : Feature screen two presentation

- Removed Model suffix from model class names
- Removed isEmpty from model Character class
- Added CharacterDetailsViewModel for character details screen
- Passed appropriate viewmodel class in ViewModelFactoryProvider.kt
- Changed dimens naming convention

anarahari committed:
feat : Feature screen two unit tests

- Added unit tests to check success and failure for presentation viewmodel layer
- Added unit tests to check success and failure for domain use case layer

anarahari committed:
feat : Feature screen three domain

- Created EpisodeRepository interface to get episode details

- Created GetEpisodeDetailsUseCase to fetch the episode details response from repository

- Added unit test cases for GetEpisodeDetailsUseCase to fetch episode details on success and failure scenarios.

- Added characters variable to Episode model to fetch list of characters.

- Changes for lintchecks

shreyansh-mm commented on app/src/main/java/com/compose/rickymortypoc/MainActivity.kt:
```diff
@@ -21,12 +18,12 @@ class MainActivity : ComponentActivity() {
+//                    ViewModelProvider(
+//                        LocalContext.current as ComponentActivity,
+//                        viewModelFactoryProvider
+//                    )[CharacterViewModel::class.java]
```
Remove the commented code

shreyansh-mm commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -1,17 +1,62 @@
+    }
+}
+
+fun mapToOriginModel(origin: GetCharacterDetailsQuery.Origin?): Origin {
```
private

shreyansh-mm commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -1,17 +1,62 @@
+    )
+}
+
+fun mapToLocationModel(location: GetCharacterDetailsQuery.Location?): Location {
```
private

shreyansh-mm commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -1,17 +1,62 @@
+    )
+}
+
+fun mapToEpisodeModel(episodes: List<GetCharacterDetailsQuery.Episode?>?): List<Episode>? {
```
private

shreyansh-mm commented on data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt:
```diff
@@ -21,4 +23,13 @@ internal class CharacterRepositoryImpl @Inject constructor(private val apolloCli
+            getCharacterDetailsQueryToCharacterModel(response.data)
+        } catch (ex: Exception) {
+            null
+        }
```
Exception should not be consumed, it should be propagated forward.
So we can show appropriate error to use.

shreyansh-mm commented on domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt:
```diff
@@ -0,0 +1,75 @@
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(testDispatcher)
```
Test should work fine without this line

shreyansh-mm commented on domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt:
```diff
@@ -0,0 +1,75 @@
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
```
this should also be not needed

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun CharacterDetailsItems(
```
private

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun DisplayCharacterDetailsData(
```
private

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            Text(
+                text = characterState.error.toString(),
+                color = MaterialTheme.colorScheme.error,
+                modifier = Modifier
```
You don't need to pass empty modifier

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+    modifier: Modifier = Modifier,
+    character: Character,
+) {
+    Column(modifier.verticalScroll(rememberScrollState())) {
```
let's move `rememberScrollState()` to a variable

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+                    .aspectRatio(1f)
+            )
+        }
+        Spacer(modifier.padding(top = MaterialTheme.dimens.paddingExtraLarge))
```
Modifier coming from args should only apply to parent composable.

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            text = character.name.orEmpty(),
+            fontWeight = FontWeight.Bold,
+            style = MaterialTheme.typography.headlineLarge,
+            color = Color.Black,
```
This should come from `MaterialTheme`

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            fontWeight = FontWeight.Bold,
+            style = MaterialTheme.typography.headlineLarge,
+            color = Color.Black,
+            modifier = modifier.padding(start = MaterialTheme.dimens.paddingExtraLarge)
```
New modifier should be used

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            color = Color.Black,
+            modifier = modifier.padding(start = MaterialTheme.dimens.paddingExtraLarge)
+        )
+        CharacterStatus(modifier, character)
```
This modifier should not be passed forward

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun CharacterStatus(modifier: Modifier, character: Character) {
```
private

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            text = stringResource(R.string.status),
+            fontSize = 16.sp,
+            fontWeight = FontWeight.Bold,
+            color = Color.Black
```
all these values can come from `textStyle` / `typography`

shreyansh-mm commented on presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt:
```diff
@@ -15,29 +16,43 @@ import androidx.compose.material3.CircularProgressIndicator
+            areCharactersFetched = true
+            viewModel.getCharacters()
+        }
+    }
```
our UI layer needs to be as dumb as possible, this logic should not sit here

anarahari committed:
feat : Feature screen three Domain

- PR review comments resolution

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun CharacterStatus(modifier: Modifier = Modifier, character: Character) {
```
This composable should only ask for the `status` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun CharacterSpecies(modifier: Modifier, character: Character) {
```
This composable should only ask for the `species` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun CharacterGender(modifier: Modifier, character: Character) {
```
This composable should only ask for the `gender` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun CharacterOrigin(modifier: Modifier, character: Character) {
```
This composable should only ask for the `origin` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun CharacterLocation(modifier: Modifier, character: Character) {
```
This composable should only ask for the `location` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+}
+
+@Composable
+private fun DisplayEpisodes(modifier: Modifier, character: Character) {
```
This composable should only ask for the `episodes` variable

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,310 @@
+                it,
+                key = { item -> character.episodes?.get(item)?.id!! }
+            ) { index ->
+                EpisodesItem(modifier, episode = character.episodes?.get(index)!!)
```
Avoid using the double bang operator (`!!`) whenever you can

anarahari committed:
feat : Feature screen three Domain

- PR review comments resolution
- passed only individual data to each composable functions
for example : CharacterStatus will have only status data, CharacterGender will have only gender data
